### PR TITLE
Add metadata to setProviderSession

### PR DESCRIPTION
**Pull request by skeltonm**
@panva This is more of a starting point so I can get your feedback on what needs to be updated or changed with the implementation of the meta argument.

Ref #352

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -354,19 +354,22 @@ async setProviderSession(req, res, {
...
+    [client_id]: {},
+  }
 } = {})
 ```
 
 ```js
 // with express
 expressApp.post('/interaction/:grant/login', async (req, res) => {
-  await provider.setProviderSession(req, res, { account: 'accountId' });
+  await provider.setProviderSession(req, res, { account: 'accountId' }, { [client_id]: { key: 'value' } });
```
Comment (panva) [inline docs/configuration.md]: add the clients array to this example then, or omit the meta from it.

File: docs/configuration.md [hunk 2]
```diff
@@ -354,19 +354,22 @@ async setProviderSession(req, res, {
...
   // ...
 });
 
 // with koa
 router.post('/interaction/:grant/login', async (ctx, next) => {
-  await provider.setProviderSession(ctx.req, ctx.res, { account: 'accountId' });
+  await provider.setProviderSession(req, res, { account: 'accountId' }, { [client_id]: { key: 'value' } });
```
Comment (panva) [inline docs/configuration.md]: dtto.

File: lib/provider.js [hunk 3]
```diff
@@ -248,6 +249,7 @@ class Provider extends events.EventEmitter {
     Object.assign(session, {
       account,
       loginTs: ts,
+      meta,
```
Comment (panva) [inline lib/provider.js]: Nope, similar to `sidFor` it should call `metaFor` with the provided client_id and meta object. It should also assert that the client_id was provided in the clients array.

File: lib/models/session.js [hunk 4]
```diff
@@ -40,6 +40,10 @@ module.exports = function getSession(provider) {
       return this.loginTs;
     }
 
+    meta() {
```
Comment (panva) [inline lib/models/session.js]: this is not necessary, there's `metaFor`

**Comment by panva** [comment]
@skeltonm thank you for proposing this, i left my comments in the review.

File: docs/configuration.md (hunk 1 above)
Comment (skeltonm) [reply docs/configuration.md]: I removed the metadata since I put it in the wrong spot anyway.

File: lib/models/session.js (hunk 4 above)
Comment (skeltonm) [reply lib/models/session.js]: Removed

File: docs/configuration.md [hunk 5]
```diff
@@ -366,7 +369,7 @@ expressApp.post('/interaction/:grant/login', async (req, res) => {
 
 // with koa
 router.post('/interaction/:grant/login', async (ctx, next) => {
-  await provider.setProviderSession(ctx.req, ctx.res, { account: 'accountId' });
+  await provider.setProviderSession(req, res, { account: 'accountId' });
```
Comment (panva) [inline docs/configuration.md]: ctx.req, ctx.res got changed to req, res

Comment (skeltonm) [reply docs/configuration.md]: Ah, good catch.  I did a copy and paste from the Express example.

Comment (skeltonm) [reply docs/configuration.md]: Updated

File: lib/provider.js [hunk 6]
```diff
@@ -249,13 +249,17 @@ class Provider extends events.EventEmitter {
...
       account,
       loginTs: ts,
-      meta,
     });
     if (!remember) session.transient = true;
     clients.forEach((clientId) => {
       assert(typeof clientId === 'string', 'clients must contain an array of client_id strings');
       session.sidFor(clientId, uuid());
     });
+    Object.keys(meta).forEach((clientId) => {
```
Comment (panva) [inline lib/provider.js]: please use `Object.entries(meta).forEach(([clientId, meta]) => { ... });` instead

**Pull request merged**